modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/QueryTransactionWrapperSelfTest.java

```diff
@@ -91,27 +97,51 @@ public void testCommitImplicit() {
 
     @Test
     public void testRollbackImplicit() {
-        QueryTransactionWrapper wrapper = prepareImplicitTx();
-        NoOpTransaction tx = (NoOpTransaction) wrapper.unwrap();
+        NoOpTransaction tx = new NoOpTransaction("test");
+        QueryTransactionWrapper wrapper = new QueryTransactionWrapper(tx, true);
 
         wrapper.rollback();
 
         assertThat(tx.rollbackFuture().isDone(), equalTo(true));
         assertThat(tx.commitFuture().isDone(), equalTo(false));
     }
 
-    private QueryTransactionWrapper prepareImplicitTx() {
-        when(transactions.begin(any())).thenReturn(new NoOpTransaction("test"));
+    @Test
```

Comment:
Let's add a test calling `startTxIfNeeded` twice, which should be impossible, because it semantic of nested transactions.

Reply:
Done

Reply:
Added test `throwsExceptionForTxControlStatementInsideExternalTransaction`.